File: backend/api/v1/schemas/track_download.py
"""Schemas for the inline track-download feature.

This is a fork-only addition. Requests are proxied to a yt-dlp-worker sidecar
on gnat which performs the actual yt-dlp call and post-download Lidarr/Plex
triggers. The library label (music | music-personal) is stamped server-side
from the MUSICSEERR_LIBRARY env var so that public musicseerr cannot drop
files into the personal library.

Search source: "youtube" (default; free-text yt-dlp search) or "spotify"
(Spotify Web API search; the worker resolves the chosen Spotify track to a
matching YouTube video at download time, transparently to the client).
"""

from __future__ import annotations

from typing import Literal

from infrastructure.msgspec_fastapi import AppStruct


SearchSource = Literal["youtube", "spotify"]


class TrackDownloadSearchRequest(AppStruct):
    query: str
    limit: int = 5
    source: SearchSource = "youtube"


class TrackDownloadCandidate(AppStruct):
    # video_id is a YouTube video ID when source="youtube" and a Spotify
    # track ID when source="spotify". The worker disambiguates by source.
    video_id: str
    url: str
    title: str
    source: SearchSource = "youtube"
    channel: str | None = None
    artist: str | None = None  # populated for source="spotify"
    album: str | None = None  # populated for source="spotify"
    duration_seconds: int | None = None
    thumbnail_url: str | None = None


class TrackDownloadSearchResponse(AppStruct):
    candidates: list[TrackDownloadCandidate]


class TrackDownloadRequest(AppStruct):
    """Inbound request from the frontend. The library param is intentionally
    NOT included here — the backend stamps it from env config so the public
    musicseerr instance cannot target the personal library."""

    video_id: str
    artist: str
    album: str
    track_title: str
    source: SearchSource = "youtube"
    target_duration_seconds: int | None = None  # passed through for spotify→yt resolution
    artist_mbid: str | None = None
    track_position: int | None = None
    disc_number: int | None = None


class TrackDownloadAccepted(AppStruct):
    job_id: str


class TrackDownloadJobStatus(AppStruct):
    id: str
    status: str
    artist: str
    album: str
    track_title: str
    library: str
    created_at: str
    updated_at: str
    file_path: str | None = None
    error: str | None = None

File: frontend/src/lib/api/trackDownload.ts
import { api } from '$lib/api/client';

export type TrackDownloadStatus =
	| 'queued'
	| 'searching'
	| 'downloading'
	| 'tagging'
	| 'importing'
	| 'done'
	| 'failed';

export type TrackDownloadSource = 'youtube' | 'spotify';

export const TRACK_DOWNLOAD_TERMINAL_STATES: ReadonlySet<TrackDownloadStatus> = new Set([
	'done',
	'failed'
]);

export interface TrackDownloadCandidate {
	video_id: string;
	url: string;
	title: string;
	source: TrackDownloadSource;
	channel: string | null;
	artist: string | null;
	album: string | null;
	duration_seconds: number | null;
	thumbnail_url: string | null;
}

export interface TrackDownloadSearchResponse {
	candidates: TrackDownloadCandidate[];
}

export interface TrackDownloadRequestPayload {
	video_id: string;
	artist: string;
	album: string;
	track_title: string;
	source?: TrackDownloadSource;
	target_duration_seconds?: number | null;
	artist_mbid?: string | null;
	track_position?: number | null;
	disc_number?: number | null;
}

export interface TrackDownloadAccepted {
	job_id: string;
}

export interface TrackDownloadJobStatus {
	id: string;
	status: TrackDownloadStatus;
	artist: string;
	album: string;
	track_title: string;
	library: string;
	created_at: string;
	updated_at: string;
	file_path: string | null;
	error: string | null;
}

const ROOT = '/api/v1/track-download';

export async function searchTrackCandidates(
	query: string,
	limit = 5,
	source: TrackDownloadSource = 'youtube',
	signal?: AbortSignal
): Promise<TrackDownloadSearchResponse> {
	return api.global.post<TrackDownloadSearchResponse>(
		`${ROOT}/search`,
		{ query, limit, source },
		{ signal }
	);
}

export async function requestTrackDownload(
	payload: TrackDownloadRequestPayload,
	signal?: AbortSignal
): Promise<TrackDownloadAccepted> {
	return api.global.post<TrackDownloadAccepted>(ROOT, payload, { signal });
}

export async function getTrackDownloadJob(
	jobId: string,
	signal?: AbortSignal
): Promise<TrackDownloadJobStatus> {
	return api.global.get<TrackDownloadJobStatus>(`${ROOT}/${jobId}`, { signal });
}
